File: langs/python/src/guidearch/models/scenario_loader.py
"""Load and validate a scenario JSON file → ScenarioM.

Path depth (from scenario_loader.py to repo root):
  scenario_loader.py → models(0) → guidearch(1) → src(2) → python(3) → langs(4) → root(5)
"""

from __future__ import annotations

import json
from pathlib import Path
from typing import Any

import jsonschema

from guidearch.models.alternative import AlternativeM
from guidearch.models.coefficient import CoefficientM
from guidearch.models.constraint import (
    ConflictConstraint,
    Constraint,
    DependencyConstraint,
    ThresholdConstraint,
)
from guidearch.models.decision import DecisionM
from guidearch.models.normalized_fuzzy import NormalizedFuzzyM
from guidearch.models.property import PropertyM
from guidearch.models.scenario import ConfigM, ScenarioM
from guidearch.models.triangular_fuzzy import TriangularFuzzyM

_REPO_ROOT = Path(__file__).parents[5]
_DEFAULT_SCHEMA_PATH = _REPO_ROOT / "spec" / "domain" / "scenario.schema.json"


class ScenarioValidationError(ValueError):
    """Raised for fatal invariant violations in a scenario."""


def load_scenario(
    path: Path,
    schema_path: Path | None = None,
) -> ScenarioM:
    """Load, validate, and parse a scenario JSON file.

    Args:
        path: Path to the scenario JSON.
        schema_path: Override the JSON Schema path (optional).

    Returns:
        A validated ScenarioM with any warning messages in `.warnings`.

    Raises:
        ScenarioValidationError: on any fatal invariant violation.
    """
    schema_path = schema_path or _DEFAULT_SCHEMA_PATH

    raw = json.loads(path.read_text(encoding="utf-8"))

    # ------------------------------------------------------------------ #
    # JSON Schema structural validation
    # ------------------------------------------------------------------ #
    schema = json.loads(schema_path.read_text(encoding="utf-8"))
    validator = jsonschema.Draft202012Validator(schema)
    errors = list(validator.iter_errors(raw))
    if errors:
        msgs = "; ".join(e.message for e in errors[:5])
        raise ScenarioValidationError(f"JSON Schema validation failed: {msgs}")

    warnings: list[str] = []

    # ------------------------------------------------------------------ #
    # Parse raw objects
    # ------------------------------------------------------------------ #
    decisions = tuple(
        DecisionM(id=d["id"], name=d["name"]) for d in raw["decisions"]
    )
    alternatives = tuple(
        AlternativeM(id=a["id"], decision_id=a["decisionId"], name=a["name"])
        for a in raw["alternatives"]
    )
    properties = tuple(
        PropertyM(id=p["id"], name=p["name"], kind=p["kind"], weight=p["weight"])
        for p in raw["properties"]
    )

    def _tfm(v: dict[str, Any]) -> TriangularFuzzyM:
        return TriangularFuzzyM(
            lower=float(v["lower"]),
            modal=float(v["modal"]),
            upper=float(v["upper"]),
        )

    coefficients = tuple(
        CoefficientM(
            alternative_id=c["alternativeId"],
            property_id=c["propertyId"],
            value=_tfm(c["value"]),
        )
        for c in raw["coefficients"]
    )

    constraints: list[Constraint] = []
    for c in raw["constraints"]:
        if c["kind"] == "threshold":
            constraints.append(
                ThresholdConstraint(
                    kind="threshold",
                    property_id=c["propertyId"],
                    min=c.get("min"),
                    max=c.get("max"),
                )
            )
        elif c["kind"] == "dependency":
            constraints.append(
                DependencyConstraint(
                    kind="dependency",
                    source_alternative_id=c["sourceAlternativeId"],
                    target_alternative_id=c["targetAlternativeId"],
                )
            )
        elif c["kind"] == "conflict":
            constraints.append(
                ConflictConstraint(
                    kind="conflict",
                    alternative_a_id=c["alternativeAId"],
                    alternative_b_id=c["alternativeBId"],
                )
            )

    w_raw = raw["config"]["weights"]
    weights = NormalizedFuzzyM(
        positive=float(w_raw["positive"]),
        average=float(w_raw["average"]),
        negative=float(w_raw["negative"]),
    )
    config = ConfigM(
        aggregation=raw["config"]["aggregation"],
        weights=weights,
    )

    # ------------------------------------------------------------------ #
    # Invariant 1: Identifier uniqueness (fatal)
    # ------------------------------------------------------------------ #
    dec_ids = [d.id for d in decisions]
    alt_ids = [a.id for a in alternatives]
    prop_ids = [p.id for p in properties]

    if len(dec_ids) != len(set(dec_ids)):
        raise ScenarioValidationError("Invariant 1.1: duplicate decision id(s)")
    if len(alt_ids) != len(set(alt_ids)):
        raise ScenarioValidationError("Invariant 1.2: duplicate alternative id(s)")
    if len(prop_ids) != len(set(prop_ids)):
        raise ScenarioValidationError("Invariant 1.3: duplicate property id(s)")

    dec_id_set = set(dec_ids)
    alt_id_set = set(alt_ids)
    prop_id_set = set(prop_ids)

    overlap_da = dec_id_set & alt_id_set
    overlap_dp = dec_id_set & prop_id_set
    overlap_ap = alt_id_set & prop_id_set
    if overlap_da or overlap_dp or overlap_ap:
        raise ScenarioValidationError(
            f"Invariant 1.4: id namespace collision: "
            f"d∩a={overlap_da}, d∩p={overlap_dp}, a∩p={overlap_ap}"
        )

    # ------------------------------------------------------------------ #
    # Invariant 2: Cross-reference validity (fatal)
    # ------------------------------------------------------------------ #
    for a in alternatives:
        if a.decision_id not in dec_id_set:
            raise ScenarioValidationError(
                f"Invariant 2.1: alternative '{a.id}' references unknown "
                f"decision '{a.decision_id}'"
            )
    for c in coefficients:
        if c.alternative_id not in alt_id_set:
            raise ScenarioValidationError(
                f"Invariant 2.2: coefficient references unknown alternative "
                f"'{c.alternative_id}'"
            )
        if c.property_id not in prop_id_set:
            raise ScenarioValidationError(
                f"Invariant 2.3: coefficient references unknown property "
                f"'{c.property_id}'"
            )
    for con in constraints:
        if isinstance(con, ThresholdConstraint):
            if con.property_id not in prop_id_set:
                raise ScenarioValidationError(
                    f"Invariant 2.4: threshold constraint references unknown "
                    f"property '{con.property_id}'"
                )
        elif isinstance(con, DependencyConstraint):
            if con.source_alternative_id not in alt_id_set:
                raise ScenarioValidationError(
                    f"Invariant 2.5: dependency constraint references unknown "
                    f"source '{con.source_alternative_id}'"
                )
            if con.target_alternative_id not in alt_id_set:
                raise ScenarioValidationError(
                    f"Invariant 2.5: dependency constraint references unknown "
                    f"target '{con.target_alternative_id}'"
                )
        elif isinstance(con, ConflictConstraint):
            if con.alternative_a_id not in alt_id_set:
                raise ScenarioValidationError(
                    f"Invariant 2.5: conflict constraint references unknown "
                    f"alternative A '{con.alternative_a_id}'"
                )
            if con.alternative_b_id not in alt_id_set:
                raise ScenarioValidationError(
                    f"Invariant 2.5: conflict constraint references unknown "
                    f"alternative B '{con.alternative_b_id}'"
                )

    # ------------------------------------------------------------------ #
    # Invariant 3: Coefficient completeness (fatal)
    # ------------------------------------------------------------------ #
    coeff_pairs = [(c.alternative_id, c.property_id) for c in coefficients]
    coeff_pair_set = set(coeff_pairs)
    if len(coeff_pairs) != len(coeff_pair_set):
        raise ScenarioValidationError(
            "Invariant 3.1: duplicate (alternativeId, propertyId) coefficient"
        )
    for a_id in alt_ids:
        for p_id in prop_ids:
            if (a_id, p_id) not in coeff_pair_set:
                raise ScenarioValidationError(
                    f"Invariant 3.1: missing coefficient for "
                    f"(alternative={a_id}, property={p_id})"
                )

    # ------------------------------------------------------------------ #
    # Invariant 4: Triangular ordering (warning)
    # ------------------------------------------------------------------ #
    for c in coefficients:
        v = c.value
        if not (v.lower <= v.modal <= v.upper):
            warnings.append(
                f"Invariant 4.1: coefficient ({c.alternative_id}, "
                f"{c.property_id}) has lower={v.lower} modal={v.modal} "
                f"upper={v.upper} — ordering violated"
            )

    # ------------------------------------------------------------------ #
    # Invariant 5: Weights (fatal)
    # ------------------------------------------------------------------ #
    w = config.weights
    for val, label in [(w.positive, "positive"), (w.average, "average"), (w.negative, "negative")]:
        if not (0.0 <= val <= 1.0):
            raise ScenarioValidationError(
                f"Invariant 5.2: weight.{label}={val} is outside [0, 1]"
            )
    w_sum = w.positive + w.average + w.negative
    if abs(w_sum - 1.0) > 1e-9:
        raise ScenarioValidationError(
            f"Invariant 5.1: weights sum to {w_sum}, expected 1.0 (tolerance 1e-9)"
        )

    # ------------------------------------------------------------------ #
    # Invariant 6: Threshold constraints (fatal)
    # ------------------------------------------------------------------ #
    for i, con in enumerate(constraints):
        if isinstance(con, ThresholdConstraint):
            if con.min is None and con.max is None:
                raise ScenarioValidationError(
                    f"Invariant 6.1: threshold constraint[{i}] has neither min nor max"
                )
            if con.min is not None and con.max is not None and con.min > con.max:
                raise ScenarioValidationError(
                    f"Invariant 6.2: threshold constraint[{i}] has min={con.min} > "
                    f"max={con.max}"
                )

    # ------------------------------------------------------------------ #
    # Invariant 7: Dependency / conflict self-edges (fatal + warning)
    # ------------------------------------------------------------------ #
    dec_of: dict[str, str] = {a.id: a.decision_id for a in alternatives}
    for i, con in enumerate(constraints):
        if isinstance(con, DependencyConstraint):
            if con.source_alternative_id == con.target_alternative_id:
                raise ScenarioValidationError(
                    f"Invariant 7.1: dependency constraint[{i}] is a self-edge"
                )
            if dec_of.get(con.source_alternative_id) == dec_of.get(
                con.target_alternative_id
            ):
                warnings.append(
                    f"Invariant 7.2: dependency constraint[{i}] connects "
                    f"alternatives of the same decision — rarely meaningful"
                )
        elif isinstance(con, ConflictConstraint):
            if con.alternative_a_id == con.alternative_b_id:
                raise ScenarioValidationError(
                    f"Invariant 7.1: conflict constraint[{i}] is a self-edge"
                )
            if dec_of.get(con.alternative_a_id) == dec_of.get(
                con.alternative_b_id
            ):
                warnings.append(
                    f"Invariant 7.2: conflict constraint[{i}] connects "
                    f"alternatives of the same decision — rarely meaningful"
                )

    # ------------------------------------------------------------------ #
    # Invariant 8: Decision occupancy (fatal)
    # ------------------------------------------------------------------ #
    alts_by_dec: dict[str, list[str]] = {d.id: [] for d in decisions}
    for a in alternatives:
        alts_by_dec[a.decision_id].append(a.id)
    for d in decisions:
        if not alts_by_dec[d.id]:
            raise ScenarioValidationError(
                f"Invariant 8.1: decision '{d.id}' ('{d.name}') has no alternatives"
            )

    return ScenarioM(
        schema_version=raw["schemaVersion"],
        name=raw["name"],
        description=raw.get("description", ""),
        decisions=decisions,
        alternatives=alternatives,
        properties=properties,
        coefficients=coefficients,
        constraints=tuple(constraints),
        config=config,
        warnings=tuple(warnings),
    )

File: langs/typescript/src/models/scenario-loader.ts
/**
 * Load and validate a scenario JSON file → ScenarioM.
 *
 * Path from this file to the repo root:
 *   scenario-loader.ts → models(0) → src(1) → typescript(2) → langs(3) → root(4)
 */
import fs from 'fs';
import path from 'path';
import { fileURLToPath } from 'url';
import Ajv from 'ajv/dist/2020.js';
import type { ScenarioM } from './scenario.js';
import type { DecisionM } from './decision.js';
import type { AlternativeM } from './alternative.js';
import type { PropertyM } from './property.js';
import { TriangularFuzzyM } from './triangular-fuzzy.js';
import type { CoefficientM } from './coefficient.js';
import type { ConstraintM } from './constraint.js';
import type { ConfigM } from './config.js';

// ---------------------------------------------------------------------------
// Custom error type
// ---------------------------------------------------------------------------
export class ScenarioValidationError extends Error {
  constructor(
    message: string,
    public readonly errors: string[],
  ) {
    super(message);
    this.name = 'ScenarioValidationError';
  }
}

// ---------------------------------------------------------------------------
// Schema path discovery
// ---------------------------------------------------------------------------
function findRepoRoot(startDir: string): string {
  let dir = startDir;
  while (true) {
    if (fs.existsSync(path.join(dir, '.git'))) return dir;
    const parent = path.dirname(dir);
    if (parent === dir) throw new Error('Could not find repo root (.git) from ' + startDir);
    dir = parent;
  }
}

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);
const _REPO_ROOT = findRepoRoot(__dirname);
const _DEFAULT_SCHEMA_PATH = path.join(_REPO_ROOT, 'spec', 'domain', 'scenario.schema.json');

// ---------------------------------------------------------------------------
// Loader
// ---------------------------------------------------------------------------
// eslint-disable-next-line @typescript-eslint/no-explicit-any
type RawJson = Record<string, any>;

export function loadScenario(jsonPath: string, schemaPath?: string): ScenarioM {
  const resolvedSchema = schemaPath ?? _DEFAULT_SCHEMA_PATH;

  const raw: RawJson = JSON.parse(fs.readFileSync(jsonPath, 'utf-8'));
  const schema: RawJson = JSON.parse(fs.readFileSync(resolvedSchema, 'utf-8'));

  // ------------------------------------------------------------------ //
  // JSON Schema structural validation via ajv
  // ------------------------------------------------------------------ //
  const ajv = new Ajv({ strict: false });
  const validate = ajv.compile(schema);
  if (!validate(raw)) {
    const msgs = (validate.errors ?? []).slice(0, 5).map((e) => `${e.instancePath} ${e.message}`);
    throw new ScenarioValidationError(`JSON Schema validation failed: ${msgs.join('; ')}`, msgs);
  }

  const warnings: string[] = [];

  // ------------------------------------------------------------------ //
  // Parse raw objects
  // ------------------------------------------------------------------ //
  const decisions: DecisionM[] = (raw['decisions'] as RawJson[]).map((d) => ({
    id: d['id'] as string,
    name: d['name'] as string,
  }));

  const alternatives: AlternativeM[] = (raw['alternatives'] as RawJson[]).map((a) => ({
    id: a['id'] as string,
    decisionId: a['decisionId'] as string,
    name: a['name'] as string,
  }));

  const properties: PropertyM[] = (raw['properties'] as RawJson[]).map((p) => ({
    id: p['id'] as string,
    name: p['name'] as string,
    kind: p['kind'] as 'min' | 'max',
    weight: p['weight'] as number,
  }));

  const coefficients: CoefficientM[] = (raw['coefficients'] as RawJson[]).map((c) => ({
    alternativeId: c['alternativeId'] as string,
    propertyId: c['propertyId'] as string,
    value: new TriangularFuzzyM(
      c['value']['lower'] as number,
      c['value']['modal'] as number,
      c['value']['upper'] as number,
    ),
  }));

  const constraints: ConstraintM[] = (raw['constraints'] as RawJson[]).map((c) => {
    if (c['kind'] === 'threshold') {
      return {
        kind: 'threshold' as const,
        propertyId: c['propertyId'] as string,
        min: c['min'] as number | undefined,
        max: c['max'] as number | undefined,
      };
    } else if (c['kind'] === 'dependency') {
      return {
        kind: 'dependency' as const,
        sourceAlternativeId: c['sourceAlternativeId'] as string,
        targetAlternativeId: c['targetAlternativeId'] as string,
      };
    } else {
      return {
        kind: 'conflict' as const,
        alternativeAId: c['alternativeAId'] as string,
        alternativeBId: c['alternativeBId'] as string,
      };
    }
  });

  const wRaw = (raw['config'] as RawJson)['weights'] as RawJson;
  const config: ConfigM = {
    aggregation: (raw['config'] as RawJson)['aggregation'] as 'sum' | 'max',
    weights: {
      positive: wRaw['positive'] as number,
      average: wRaw['average'] as number,
      negative: wRaw['negative'] as number,
    },
  };

  // ------------------------------------------------------------------ //
  // Invariant 1: Identifier uniqueness (fatal)
  // ------------------------------------------------------------------ //
  const decIds = decisions.map((d) => d.id);
  const altIds = alternatives.map((a) => a.id);
  const propIds = properties.map((p) => p.id);

  if (new Set(decIds).size !== decIds.length) {
    throw new ScenarioValidationError('Invariant 1.1: duplicate decision id(s)', [
      'duplicate decision id(s)',
    ]);
  }
  if (new Set(altIds).size !== altIds.length) {
    throw new ScenarioValidationError('Invariant 1.2: duplicate alternative id(s)', [
      'duplicate alternative id(s)',
    ]);
  }
  if (new Set(propIds).size !== propIds.length) {
    throw new ScenarioValidationError('Invariant 1.3: duplicate property id(s)', [
      'duplicate property id(s)',
    ]);
  }

  const decIdSet = new Set(decIds);
  const altIdSet = new Set(altIds);
  const propIdSet = new Set(propIds);

  const overlapDA = decIds.filter((id) => altIdSet.has(id));
  const overlapDP = decIds.filter((id) => propIdSet.has(id));
  const overlapAP = altIds.filter((id) => propIdSet.has(id));
  if (overlapDA.length || overlapDP.length || overlapAP.length) {
    throw new ScenarioValidationError(
      `Invariant 1.4: id namespace collision: d∩a=${JSON.stringify(overlapDA)}, d∩p=${JSON.stringify(overlapDP)}, a∩p=${JSON.stringify(overlapAP)}`,
      ['id namespace collision'],
    );
  }

  // ------------------------------------------------------------------ //
  // Invariant 2: Cross-reference validity (fatal)
  // ------------------------------------------------------------------ //
  for (const a of alternatives) {
    if (!decIdSet.has(a.decisionId)) {
      throw new ScenarioValidationError(
        `Invariant 2.1: alternative '${a.id}' references unknown decision '${a.decisionId}'`,
        [`alternative '${a.id}' references unknown decision '${a.decisionId}'`],
      );
    }
  }
  for (const c of coefficients) {
    if (!altIdSet.has(c.alternativeId)) {
      throw new ScenarioValidationError(
        `Invariant 2.2: coefficient references unknown alternative '${c.alternativeId}'`,
        [`coefficient references unknown alternative '${c.alternativeId}'`],
      );
    }
    if (!propIdSet.has(c.propertyId)) {
      throw new ScenarioValidationError(
        `Invariant 2.3: coefficient references unknown property '${c.propertyId}'`,
        [`coefficient references unknown property '${c.propertyId}'`],
      );
    }
  }
  for (const con of constraints) {
    if (con.kind === 'threshold') {
      if (!propIdSet.has(con.propertyId)) {
        throw new ScenarioValidationError(
          `Invariant 2.4: threshold constraint references unknown property '${con.propertyId}'`,
          [`threshold constraint references unknown property '${con.propertyId}'`],
        );
      }
    } else if (con.kind === 'dependency') {
      if (!altIdSet.has(con.sourceAlternativeId)) {
        throw new ScenarioValidationError(
          `Invariant 2.5: dependency constraint references unknown source '${con.sourceAlternativeId}'`,
          [`dependency constraint references unknown source '${con.sourceAlternativeId}'`],
        );
      }
      if (!altIdSet.has(con.targetAlternativeId)) {
        throw new ScenarioValidationError(
          `Invariant 2.5: dependency constraint references unknown target '${con.targetAlternativeId}'`,
          [`dependency constraint references unknown target '${con.targetAlternativeId}'`],
        );
      }
    } else {
      // conflict
      if (!altIdSet.has(con.alternativeAId)) {
        throw new ScenarioValidationError(
          `Invariant 2.5: conflict constraint references unknown alternative A '${con.alternativeAId}'`,
          [`conflict constraint references unknown alternative A '${con.alternativeAId}'`],
        );
      }
      if (!altIdSet.has(con.alternativeBId)) {
        throw new ScenarioValidationError(
          `Invariant 2.5: conflict constraint references unknown alternative B '${con.alternativeBId}'`,
          [`conflict constraint references unknown alternative B '${con.alternativeBId}'`],
        );
      }
    }
  }

  // ------------------------------------------------------------------ //
  // Invariant 3: Coefficient completeness (fatal)
  // ------------------------------------------------------------------ //
  const coeffPairs = coefficients.map((c) => `${c.alternativeId}|${c.propertyId}`);
  const coeffPairSet = new Set(coeffPairs);
  if (coeffPairs.length !== coeffPairSet.size) {
    throw new ScenarioValidationError(
      'Invariant 3.1: duplicate (alternativeId, propertyId) coefficient',
      ['duplicate (alternativeId, propertyId) coefficient'],
    );
  }
  for (const aId of altIds) {
    for (const pId of propIds) {
      if (!coeffPairSet.has(`${aId}|${pId}`)) {
        throw new ScenarioValidationError(
          `Invariant 3.1: missing coefficient for (alternative=${aId}, property=${pId})`,
          [`missing coefficient for (alternative=${aId}, property=${pId})`],
        );
      }
    }
  }

  // ------------------------------------------------------------------ //
  // Invariant 4: Triangular ordering (warning)
  // ------------------------------------------------------------------ //
  for (const c of coefficients) {
    const v = c.value;
    if (!(v.lower <= v.modal && v.modal <= v.upper)) {
      warnings.push(
        `Invariant 4.1: coefficient (${c.alternativeId}, ${c.propertyId}) has lower=${v.lower} modal=${v.modal} upper=${v.upper} — ordering violated`,
      );
    }
  }

  // ------------------------------------------------------------------ //
  // Invariant 5: Weights (fatal)
  // ------------------------------------------------------------------ //
  const w = config.weights;
  for (const [val, label] of [
    [w.positive, 'positive'],
    [w.average, 'average'],
    [w.negative, 'negative'],
  ] as [number, string][]) {
    if (val < 0 || val > 1) {
      throw new ScenarioValidationError(`Invariant 5.2: weight.${label}=${val} is outside [0, 1]`, [
        `weight.${label}=${val} is outside [0, 1]`,
      ]);
    }
  }
  const wSum = w.positive + w.average + w.negative;
  if (Math.abs(wSum - 1.0) > 1e-9) {
    throw new ScenarioValidationError(
      `Invariant 5.1: weights sum to ${wSum}, expected 1.0 (tolerance 1e-9)`,
      [`weights sum to ${wSum}, expected 1.0`],
    );
  }

  // ------------------------------------------------------------------ //
  // Invariant 6: Threshold constraints (fatal)
  // ------------------------------------------------------------------ //
  for (let i = 0; i < constraints.length; i++) {
    const con = constraints[i];
    if (con.kind === 'threshold') {
      if (con.min === undefined && con.max === undefined) {
        throw new ScenarioValidationError(
          `Invariant 6.1: threshold constraint[${i}] has neither min nor max`,
          [`threshold constraint[${i}] has neither min nor max`],
        );
      }
      if (con.min !== undefined && con.max !== undefined && con.min > con.max) {
        throw new ScenarioValidationError(
          `Invariant 6.2: threshold constraint[${i}] has min=${con.min} > max=${con.max}`,
          [`threshold constraint[${i}] has min=${con.min} > max=${con.max}`],
        );
      }
    }
  }

  // ------------------------------------------------------------------ //
  // Invariant 7: Dependency / conflict self-edges (fatal + warning)
  // ------------------------------------------------------------------ //
  const decOf: Map<string, string> = new Map(alternatives.map((a) => [a.id, a.decisionId]));
  for (let i = 0; i < constraints.length; i++) {
    const con = constraints[i];
    if (con.kind === 'dependency') {
      if (con.sourceAlternativeId === con.targetAlternativeId) {
        throw new ScenarioValidationError(
          `Invariant 7.1: dependency constraint[${i}] is a self-edge`,
          [`dependency constraint[${i}] is a self-edge`],
        );
      }
      if (decOf.get(con.sourceAlternativeId) === decOf.get(con.targetAlternativeId)) {
        warnings.push(
          `Invariant 7.2: dependency constraint[${i}] connects alternatives of the same decision — rarely meaningful`,
        );
      }
    } else if (con.kind === 'conflict') {
      if (con.alternativeAId === con.alternativeBId) {
        throw new ScenarioValidationError(
          `Invariant 7.1: conflict constraint[${i}] is a self-edge`,
          [`conflict constraint[${i}] is a self-edge`],
        );
      }
      if (decOf.get(con.alternativeAId) === decOf.get(con.alternativeBId)) {
        warnings.push(
          `Invariant 7.2: conflict constraint[${i}] connects alternatives of the same decision — rarely meaningful`,
        );
      }
    }
  }

  // ------------------------------------------------------------------ //
  // Invariant 8: Decision occupancy (fatal)
  // ------------------------------------------------------------------ //
  const altsByDec: Map<string, string[]> = new Map(decisions.map((d) => [d.id, []]));
  for (const a of alternatives) {
    altsByDec.get(a.decisionId)!.push(a.id);
  }
  for (const d of decisions) {
    if (!altsByDec.get(d.id)?.length) {
      throw new ScenarioValidationError(
        `Invariant 8.1: decision '${d.id}' ('${d.name}') has no alternatives`,
        [`decision '${d.id}' has no alternatives`],
      );
    }
  }

  return {
    schemaVersion: raw['schemaVersion'] as string,
    name: raw['name'] as string,
    description: (raw['description'] as string | undefined) ?? '',
    decisions,
    alternatives,
    properties,
    coefficients,
    constraints,
    config,
    warnings,
  };
}
